File: cli/src/catalog.ts
// ---------------------------------------------------------------------------
// Template catalog — loads agents, skills, and teams from YAML files
// ---------------------------------------------------------------------------

import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import YAML from "yaml";
import type { Skill, TeamDefinition, TeamMember } from "./adapters/base.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const TEMPLATES_DIR = path.resolve(__dirname, "../../templates");

// ---------------------------------------------------------------------------
// Catalog file types
// ---------------------------------------------------------------------------

interface CatalogAgent {
  name: string;
  role: string;
  category: string;
  soul: string;
}

interface CatalogSkill {
  id: string;
  title: string;
  category: string;
  description?: string;
  alwaysApply?: boolean;
  globs?: string[];
  userInvocable?: boolean;
  body: string;
}

interface CatalogTeam {
  label: string;
  description: string;
  defaultPlatforms: string[];
  name: string;
  agents: string[];
  skills: string[];
  agentSkills?: Record<string, string[]>;
}

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

export interface AgentCategory {
  id: string;
  label: string;
  agents: string[];
}

export interface SkillCategory {
  id: string;
  label: string;
  skills: string[];
}

export interface TeamTemplate {
  id: string;
  label: string;
  description: string;
  defaultPlatforms: string[];
  teamName: string;
  members: TeamMember[];
  skills: Skill[];
}

// ---------------------------------------------------------------------------
// Loaders
// ---------------------------------------------------------------------------

function readYaml<T>(filePath: string): T {
  return YAML.parse(fs.readFileSync(filePath, "utf-8")) as T;
}

const SAFE_NAME_RE = /^[a-zA-Z0-9][a-zA-Z0-9_-]*$/;

export function loadAgent(name: string): CatalogAgent {
  if (!SAFE_NAME_RE.test(name)) throw new Error(`Invalid agent name in template: ${JSON.stringify(name)}`);
  return readYaml<CatalogAgent>(path.join(TEMPLATES_DIR, "agents", `${name}.yaml`));
}

export function loadSkill(id: string): CatalogSkill {
  if (!SAFE_NAME_RE.test(id)) throw new Error(`Invalid skill ID in template: ${JSON.stringify(id)}`);
  return readYaml<CatalogSkill>(path.join(TEMPLATES_DIR, "skills", `${id}.yaml`));
}

function loadTeamRaw(id: string): CatalogTeam {
  return readYaml<CatalogTeam>(path.join(TEMPLATES_DIR, "teams", `${id}.yaml`));
}

// ---------------------------------------------------------------------------
// Directory scanning helpers
// ---------------------------------------------------------------------------

function scanYamlDir(dir: string): string[] {
  if (!fs.existsSync(dir)) return [];
  return fs
    .readdirSync(dir)
    .filter((f) => f.endsWith(".yaml") && f !== "_index.yaml")
    .map((f) => f.replace(/\.yaml$/, ""));
}

function tryReadIndex<T>(dir: string): T | null {
  const indexPath = path.join(dir, "_index.yaml");
  if (!fs.existsSync(indexPath)) return null;
  return readYaml<T>(indexPath);
}

// ---------------------------------------------------------------------------
// Index readers (auto-discover files, use _index.yaml for ordering/labels)
// ---------------------------------------------------------------------------

export function listTeams(): string[] {
  const dir = path.join(TEMPLATES_DIR, "teams");
  const index = tryReadIndex<{ order: string[] }>(dir);
  const onDisk = scanYamlDir(dir);
  if (!index) return onDisk;

  // Start with index order, append any new files not in the index
  const ordered = index.order.filter((id) => onDisk.includes(id));
  for (const id of onDisk) {
    if (!ordered.includes(id)) ordered.push(id);
  }
  return ordered;
}

export function listAgentCategories(): AgentCategory[] {
  const dir = path.join(TEMPLATES_DIR, "agents");
  const index = tryReadIndex<{ categories: AgentCategory[] }>(dir);
  const onDisk = scanYamlDir(dir);

  // Track which agents are already categorized
  const categorized = new Set<string>();
  const categories: AgentCategory[] = [];

  if (index) {
    for (const cat of index.categories) {
      // Only include agents that exist on disk
      const agents = cat.agents.filter((a) => onDisk.includes(a));
      agents.forEach((a) => categorized.add(a));
      categories.push({ ...cat, agents });
    }
  }

  // Auto-discover agents not in the index, group by their file's category field
  const uncategorized: Record<string, string[]> = {};
  for (const name of onDisk) {
    if (categorized.has(name)) continue;
    try {
      const agent = loadAgent(name);
      const cat = agent.category || "uncategorized";
      if (!uncategorized[cat]) uncategorized[cat] = [];
      uncategorized[cat].push(name);
    } catch {
      // Skip files that can't be parsed
    }
  }

  // Merge discovered agents into existing categories or create new ones
  for (const [catId, agents] of Object.entries(uncategorized)) {
    const existing = categories.find((c) => c.id === catId);
    if (existing) {
      existing.agents.push(...agents);
    } else {
      categories.push({ id: catId, label: catId, agents });
    }
  }

  return categories;
}

export function listSkillCategories(): SkillCategory[] {
  const dir = path.join(TEMPLATES_DIR, "skills");
  const index = tryReadIndex<{ categories: SkillCategory[] }>(dir);
  const onDisk = scanYamlDir(dir);

  const categorized = new Set<string>();
  const categories: SkillCategory[] = [];

  if (index) {
    for (const cat of index.categories) {
      const skills = cat.skills.filter((s) => onDisk.includes(s));
      skills.forEach((s) => categorized.add(s));
      categories.push({ ...cat, skills });
    }
  }

  // Auto-discover skills not in the index
  const uncategorized: Record<string, string[]> = {};
  for (const id of onDisk) {
    if (categorized.has(id)) continue;
    try {
      const skill = loadSkill(id);
      const cat = skill.category || "uncategorized";
      if (!uncategorized[cat]) uncategorized[cat] = [];
      uncategorized[cat].push(id);
    } catch {
      // Skip files that can't be parsed
    }
  }

  for (const [catId, skills] of Object.entries(uncategorized)) {
    const existing = categories.find((c) => c.id === catId);
    if (existing) {
      existing.skills.push(...skills);
    } else {
      categories.push({ id: catId, label: catId, skills });
    }
  }

  return categories;
}

// ---------------------------------------------------------------------------
// Resolution
// ---------------------------------------------------------------------------

/** Resolve a team template into a fully self-contained TeamTemplate */
export function resolveTeam(teamId: string): TeamTemplate {
  const team = loadTeamRaw(teamId);

  const members: TeamMember[] = team.agents.map((agentName) => {
    const agent = loadAgent(agentName);
    const agentSkillIds = team.agentSkills?.[agentName];
    return {
      name: agent.name,
      role: agent.role,
      soul: agent.soul,
      ...(agentSkillIds && agentSkillIds.length > 0 ? { skills: agentSkillIds } : {}),
    };
  });

  const skills: Skill[] = team.skills.map((skillId) => {
    const skill = loadSkill(skillId);
    return {
      id: skill.id,
      title: skill.title,
      ...(skill.description ? { description: skill.description } : {}),
      ...(skill.alwaysApply !== undefined ? { alwaysApply: skill.alwaysApply } : {}),
      ...(skill.globs ? { globs: skill.globs } : {}),
      ...(skill.userInvocable !== undefined ? { userInvocable: skill.userInvocable } : {}),
      body: skill.body,
    };
  });

  return {
    id: teamId,
    label: team.label,
    description: team.description,
    defaultPlatforms: team.defaultPlatforms,
    teamName: team.name,
    members,
    skills,
  };
}

/** Convert a resolved TeamTemplate into a TeamDefinition (for writing to .teamrc.yaml) */
export function templateToTeamDefinition(template: TeamTemplate, teamName: string): TeamDefinition {
  return {
    name: teamName,
    members: template.members.map((m) => ({
      name: m.name,
      role: m.role,
      soul: m.soul,
      ...(m.skills && m.skills.length > 0 ? { skills: m.skills } : {}),
    })),
    skills: template.skills.length > 0 ? template.skills : undefined,
  };
}

File: teamrc/lib/teamrc/catalog.ex
defmodule Teamrc.Catalog do
  @moduledoc "Loads team templates from the YAML catalog. Caches in ETS."

  @templates_dir Path.expand("../../../templates", __DIR__)
  @ets_table :trc_catalog_cache

  # --- Client API ---

  def templates_dir, do: @templates_dir

  @doc "Reload all templates from disk into ETS. Useful for development."
  def reload do
    ensure_table()
    :ets.delete_all_objects(@ets_table)
    populate_cache()
    :ok
  end

  @doc "List team template IDs in display order. Auto-discovers new team files."
  def list_teams do
    case ets_get(:list_teams) do
      {:ok, val} -> val
      :miss -> compute_and_cache(:list_teams, &do_list_teams/0)
    end
  end

  @doc "Load a team template's raw metadata (without resolving agent/skill refs)."
  def load_team_raw(id) do
    validate_id!(id)

    case ets_get({:team, id}) do
      {:ok, val} -> val
      :miss -> compute_and_cache({:team, id}, fn -> do_read_yaml!(Path.join(@templates_dir, "teams/#{id}.yaml")) end)
    end
  end

  @doc "Load an agent definition from the catalog."
  def load_agent(name) do
    validate_id!(name)

    case ets_get({:agent, name}) do
      {:ok, val} -> val
      :miss -> compute_and_cache({:agent, name}, fn -> do_read_yaml!(Path.join(@templates_dir, "agents/#{name}.yaml")) end)
    end
  end

  @doc "Load a skill definition from the catalog."
  def load_skill(id) do
    validate_id!(id)

    case ets_get({:skill, id}) do
      {:ok, val} -> val
      :miss -> compute_and_cache({:skill, id}, fn -> do_read_yaml!(Path.join(@templates_dir, "skills/#{id}.yaml")) end)
    end
  end

  @doc "List agent categories with their agent lists. Auto-discovers new agents."
  def list_agent_categories do
    case ets_get(:list_agent_categories) do
      {:ok, val} -> val
      :miss -> compute_and_cache(:list_agent_categories, &do_list_agent_categories/0)
    end
  end

  @doc "Returns the union of skill IDs recommended for an agent across all team templates."
  def agent_recommended_skills(agent_name) do
    case ets_get({:agent_recommended_skills, agent_name}) do
      {:ok, val} -> val
      :miss -> compute_and_cache({:agent_recommended_skills, agent_name}, fn -> do_agent_recommended_skills(agent_name) end)
    end
  end

  @doc "List skill categories with their skill lists. Auto-discovers new skills."
  def list_skill_categories do
    case ets_get(:list_skill_categories) do
      {:ok, val} -> val
      :miss -> compute_and_cache(:list_skill_categories, &do_list_skill_categories/0)
    end
  end

  @doc """
  Resolve a team template into a fully hydrated map with inline agent souls and skill bodies.
  Returns a map matching the shape expected by TeamLive.
  """
  def resolve_team(team_id) do
    case ets_get({:resolve_team, team_id}) do
      {:ok, val} -> val
      :miss -> compute_and_cache({:resolve_team, team_id}, fn -> do_resolve_team(team_id) end)
    end
  end

  # --- Private: ETS helpers ---

  defp ensure_table do
    case :ets.whereis(@ets_table) do
      :undefined ->
        :ets.new(@ets_table, [:named_table, :public, :set, read_concurrency: true])

      ref ->
        ref
    end
  end

  defp ets_get(key) do
    case :ets.whereis(@ets_table) do
      :undefined -> :miss
      _ref ->
        case :ets.lookup(@ets_table, key) do
          [{^key, val}] -> {:ok, val}
          [] -> :miss
        end
    end
  end

  defp compute_and_cache(key, fun) do
    ensure_table()
    val = fun.()
    :ets.insert(@ets_table, {key, val})
    val
  end

  defp populate_cache do
    # Pre-populate the aggregate queries. Individual items are loaded on-demand.
    teams = do_list_teams()
    :ets.insert(@ets_table, {:list_teams, teams})

    # Pre-load all team/agent/skill YAML files
    Enum.each(teams, fn id ->
      team = do_read_yaml!(Path.join(@templates_dir, "teams/#{id}.yaml"))
      :ets.insert(@ets_table, {{:team, id}, team})
    end)

    agents_dir = Path.join(@templates_dir, "agents")
    for name <- scan_yaml_dir(agents_dir) do
      agent = do_read_yaml!(Path.join(agents_dir, "#{name}.yaml"))
      :ets.insert(@ets_table, {{:agent, name}, agent})
    end

    skills_dir = Path.join(@templates_dir, "skills")
    for id <- scan_yaml_dir(skills_dir) do
      skill = do_read_yaml!(Path.join(skills_dir, "#{id}.yaml"))
      :ets.insert(@ets_table, {{:skill, id}, skill})
    end

    # Pre-compute derived results
    agent_cats = do_list_agent_categories()
    :ets.insert(@ets_table, {:list_agent_categories, agent_cats})

    skill_cats = do_list_skill_categories()
    :ets.insert(@ets_table, {:list_skill_categories, skill_cats})

    Enum.each(teams, fn id ->
      resolved = do_resolve_team(id)
      :ets.insert(@ets_table, {{:resolve_team, id}, resolved})
    end)
  rescue
    e ->
      require Logger
      Logger.warning("Catalog cache population failed: #{inspect(e)}")
  end

  # --- Private: validation ---

  defp validate_id!(id) do
    unless Regex.match?(~r/^[a-zA-Z0-9][a-zA-Z0-9_-]{0,63}$/, id) do
      raise ArgumentError, "invalid template id: #{inspect(id)}"
    end

    id
  end

  # --- Private: computation (reads from disk) ---

  defp do_list_teams do
    on_disk = scan_yaml_dir(Path.join(@templates_dir, "teams"))

    case do_read_yaml_safe(Path.join(@templates_dir, "teams/_index.yaml")) do
      {:ok, index} ->
        ordered = (index["order"] || []) |> Enum.filter(&(&1 in on_disk))
        new = on_disk -- ordered
        ordered ++ new

      :error ->
        on_disk
    end
  end

  defp do_list_agent_categories do
    dir = Path.join(@templates_dir, "agents")
    on_disk = scan_yaml_dir(dir) |> MapSet.new()

    categories =
      case do_read_yaml_safe(Path.join(dir, "_index.yaml")) do
        {:ok, index} ->
          (index["categories"] || [])
          |> Enum.map(fn cat ->
            Map.update!(cat, "agents", fn agents ->
              Enum.filter(agents, &MapSet.member?(on_disk, &1))
            end)
          end)

        :error ->
          []
      end

    categorized =
      categories
      |> Enum.flat_map(& &1["agents"])
      |> MapSet.new()

    uncategorized =
      on_disk
      |> MapSet.difference(categorized)
      |> Enum.reduce(%{}, fn name, acc ->
        case do_read_yaml_safe(Path.join(dir, "#{name}.yaml")) do
          {:ok, agent} ->
            cat = agent["category"] || "uncategorized"
            Map.update(acc, cat, [name], &[name | &1])

          :error ->
            acc
        end
      end)

    Enum.reduce(uncategorized, categories, fn {cat_id, agents}, cats ->
      case Enum.find_index(cats, &(&1["id"] == cat_id)) do
        nil ->
          cats ++ [%{"id" => cat_id, "label" => cat_id, "agents" => Enum.reverse(agents)}]

        idx ->
          List.update_at(cats, idx, fn cat ->
            Map.update!(cat, "agents", &(&1 ++ Enum.reverse(agents)))
          end)
      end
    end)
  end

  defp do_agent_recommended_skills(agent_name) do
    do_list_teams()
    |> Enum.reduce(MapSet.new(), fn team_id, acc ->
      team = load_team_raw(team_id)
      agent_skills = team["agentSkills"] || %{}
      skills = Map.get(agent_skills, agent_name, [])
      MapSet.union(acc, MapSet.new(skills))
    end)
    |> MapSet.to_list()
    |> Enum.sort()
  end

  defp do_list_skill_categories do
    dir = Path.join(@templates_dir, "skills")
    on_disk = scan_yaml_dir(dir) |> MapSet.new()

    categories =
      case do_read_yaml_safe(Path.join(dir, "_index.yaml")) do
        {:ok, index} ->
          (index["categories"] || [])
          |> Enum.map(fn cat ->
            Map.update!(cat, "skills", fn skills ->
              Enum.filter(skills, &MapSet.member?(on_disk, &1))
            end)
          end)

        :error ->
          []
      end

    categorized =
      categories
      |> Enum.flat_map(& &1["skills"])
      |> MapSet.new()

    uncategorized =
      on_disk
      |> MapSet.difference(categorized)
      |> Enum.reduce(%{}, fn id, acc ->
        case do_read_yaml_safe(Path.join(dir, "#{id}.yaml")) do
          {:ok, skill} ->
            cat = skill["category"] || "uncategorized"
            Map.update(acc, cat, [id], &[id | &1])

          :error ->
            acc
        end
      end)

    Enum.reduce(uncategorized, categories, fn {cat_id, skills}, cats ->
      case Enum.find_index(cats, &(&1["id"] == cat_id)) do
        nil ->
          cats ++ [%{"id" => cat_id, "label" => cat_id, "skills" => Enum.reverse(skills)}]

        idx ->
          List.update_at(cats, idx, fn cat ->
            Map.update!(cat, "skills", &(&1 ++ Enum.reverse(skills)))
          end)
      end
    end)
  end

  defp do_resolve_team(team_id) do
    team = load_team_raw(team_id)
    agent_skills = team["agentSkills"] || %{}

    members =
      (team["agents"] || [])
      |> Enum.map(fn agent_name ->
        agent = load_agent(agent_name)
        skills = Map.get(agent_skills, agent_name, [])

        base = %{
          name: agent["name"],
          role: agent["role"],
          soul: agent["soul"] || ""
        }

        base = if agent["description"], do: Map.put(base, :description, agent["description"]), else: base
        if skills != [], do: Map.put(base, :skills, skills), else: base
      end)

    skills =
      (team["skills"] || [])
      |> Enum.map(fn skill_id ->
        skill = load_skill(skill_id)

        base = %{id: skill["id"]}
        base = if skill["title"], do: Map.put(base, :title, skill["title"]), else: base
        base = if skill["description"], do: Map.put(base, :description, skill["description"]), else: base
        base = if skill["alwaysApply"], do: Map.put(base, :alwaysApply, true), else: base
        base = if skill["globs"], do: Map.put(base, :globs, skill["globs"]), else: base
        base = if skill["body"], do: Map.put(base, :body, skill["body"]), else: base
        base
      end)

    %{
      label: team["label"],
      description: team["description"],
      icon: team_icon(team_id),
      team_name: team["name"],
      default_platforms: team["defaultPlatforms"] || [],
      members: members,
      skills: skills
    }
  end

  # Map team IDs to icons for the web UI
  defp team_icon("fullstack"), do: "code"
  defp team_icon("backend"), do: "server"
  defp team_icon("security"), do: "shield"
  defp team_icon("marketing"), do: "megaphone"
  defp team_icon("research"), do: "book"
  defp team_icon("devops"), do: "cloud"
  defp team_icon("custom"), do: "wrench"
  defp team_icon(_), do: "code"

  defp scan_yaml_dir(dir) do
    case File.ls(dir) do
      {:ok, files} ->
        files
        |> Enum.filter(&(String.ends_with?(&1, ".yaml") and &1 != "_index.yaml"))
        |> Enum.map(&String.replace_suffix(&1, ".yaml", ""))
        |> Enum.sort()

      {:error, _} ->
        []
    end
  end

  defp do_read_yaml!(path) do
    path
    |> File.read!()
    |> YamlElixir.read_from_string!()
  end

  defp do_read_yaml_safe(path) do
    case File.read(path) do
      {:ok, content} -> {:ok, YamlElixir.read_from_string!(content)}
      {:error, _} -> :error
    end
  end
end
